src/mesh/mesh_base.C

```diff
@@ -478,45 +476,70 @@ unsigned int MeshBase::recalculate_n_partitions()
 #ifdef LIBMESH_ENABLE_DEPRECATED
 const PointLocatorBase & MeshBase::point_locator () const
 {
-  libmesh_deprecated();
+  auto point_locator_ptr = _point_locator->load(std::memory_order_acquire);
 
-  if (_point_locator.get() == nullptr)
+  // If there's no master point locator, then we need one.
+  if (point_locator_ptr == nullptr)
     {
-      // PointLocator construction may not be safe within threads
-      libmesh_assert(!Threads::in_threads);
+      Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
+
+      if (point_locator_ptr == nullptr)
+        {
+          // And it may require parallel communication
+          parallel_object_only();
```

Comment:
If we do things this way we pretty much have to switch in libmesh.C from MPI_THREAD_FUNNELED to MPI_THREAD_SERIALIZED or MPI_THREAD_MULTIPLE as per the standard.  There's no good way for us to guarantee that the thread which ends up building the point locator is the same as the thread which called MPI_Init_thread.

Reply:
@friedmud, based on the above and on the discussion last week, should we close this PR?  If you're compiling some MPI implementation(s) without threads to improve performance then there's no way we can safely risk making MPI calls from non-main threads.